File: src/com/mistraltech/smogen/codegenerator/matchergenerator/MatcherGeneratorProperties.java
package com.mistraltech.smogen.codegenerator.matchergenerator;

import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiClassType;
import com.mistraltech.smogen.codegenerator.CodeWriter;
import com.mistraltech.smogen.codegenerator.JavaGeneratorProperties;

public class MatcherGeneratorProperties extends JavaGeneratorProperties<MatcherGeneratorProperties> {
    private String factoryMethodPrefix;
    private boolean extensible;
    private String matcherSuperClassName;
    private CodeWriter matcherGeneratorCodeWriter;
    private PsiClass sourceClass;
    private String baseClassName;
    private String factoryMethodSuffix;
    private String templateFactoryMethodSuffix;
    private String setterPrefix;
    private String setterSuffix;
    private boolean useReflectingPropertyMatcher;
    private boolean generateTemplateFactoryMethod;
    private boolean makeMethodParametersFinal;
    private boolean generateInterface;

    public MatcherGeneratorProperties() {
    }

    public String getFactoryMethodPrefix() {
        return factoryMethodPrefix;
    }

    public MatcherGeneratorProperties setFactoryMethodPrefix(String factoryMethodPrefix) {
        this.factoryMethodPrefix = factoryMethodPrefix;
        return self();
    }

    public boolean isExtensible() {
        return extensible;
    }

    public MatcherGeneratorProperties setExtensible(boolean extensible) {
        this.extensible = extensible;
        return self();
    }

    public String getMatcherSuperClassName() {
        return matcherSuperClassName;
    }

    public MatcherGeneratorProperties setMatcherSuperClassName(String matcherSuperClassName) {
        this.matcherSuperClassName = matcherSuperClassName;
        return self();
    }

    public PsiClass getSourceClass() {
        return sourceClass;
    }

    public MatcherGeneratorProperties setSourceClass(PsiClass sourceClass) {
        this.sourceClass = sourceClass;
        return self();
    }

    public PsiClassType getSourceSuperClassType() {
        final PsiClassType[] extendsListTypes = sourceClass.getExtendsListTypes();
        if (extendsListTypes.length == 0) {
            return null;
        } else {
            return extendsListTypes[0];
        }
    }

    @Override
    public CodeWriter getCodeWriter() {
        return isGenerateInterface() ? new MatcherInterfaceCodeWriter(this) : new MatcherClassCodeWriter(this);
    }

    public String getBaseClassName() {
        return baseClassName;
    }

    public MatcherGeneratorProperties setBaseClassName(String baseClassName) {
        this.baseClassName = baseClassName;
        return self();
    }

    public String getFactoryMethodSuffix() {
        return factoryMethodSuffix;
    }

    public MatcherGeneratorProperties setFactoryMethodSuffix(String factoryMethodSuffix) {
        this.factoryMethodSuffix = factoryMethodSuffix;
        return self();
    }

    public String getTemplateFactoryMethodSuffix() {
        return templateFactoryMethodSuffix;
    }

    public MatcherGeneratorProperties setTemplateFactoryMethodSuffix(String templateFactoryMethodSuffix) {
        this.templateFactoryMethodSuffix = templateFactoryMethodSuffix;
        return self();
    }

    public String getSetterPrefix() {
        return setterPrefix;
    }

    public MatcherGeneratorProperties setSetterPrefix(String setterPrefix) {
        this.setterPrefix = setterPrefix;
        return self();
    }

    public String getSetterSuffix() {
        return setterSuffix;
    }

    public MatcherGeneratorProperties setSetterSuffix(String setterSuffix) {
        this.setterSuffix = setterSuffix;
        return self();
    }

    public boolean isUseReflectingPropertyMatcher() {
        return useReflectingPropertyMatcher;
    }

    public MatcherGeneratorProperties setUseReflectingPropertyMatcher(boolean useReflectingPropertyMatcher) {
        this.useReflectingPropertyMatcher = useReflectingPropertyMatcher;
        return self();
    }

    public boolean isGenerateTemplateFactoryMethod() {
        return generateTemplateFactoryMethod;
    }

    public MatcherGeneratorProperties setGenerateTemplateFactoryMethod(boolean generateTemplateFactoryMethod) {
        this.generateTemplateFactoryMethod = generateTemplateFactoryMethod;
        return self();
    }

    public boolean isMakeMethodParametersFinal() {
        return makeMethodParametersFinal;
    }

    public MatcherGeneratorProperties setMakeMethodParametersFinal(boolean makeMethodParametersFinal) {
        this.makeMethodParametersFinal = makeMethodParametersFinal;
        return self();
    }

    public boolean isGenerateInterface() {
        return generateInterface;
    }

    public MatcherGeneratorProperties setGenerateInterface(boolean generateInterface) {
        this.generateInterface = generateInterface;
        return self();
    }
}

File: src/main/kotlin/com/mistraltech/smogen/codegenerator/matchergenerator/MatcherGeneratorProperties.kt
package com.mistraltech.smogen.codegenerator.matchergenerator

import com.intellij.psi.PsiClass
import com.intellij.psi.PsiClassType
import com.mistraltech.smogen.codegenerator.CodeWriter
import com.mistraltech.smogen.codegenerator.JavaGeneratorProperties

@Suppress("TooManyFunctions")
class MatcherGeneratorProperties : JavaGeneratorProperties<MatcherGeneratorProperties>() {
    var factoryMethodPrefix: String? = null
        private set
    var isExtensible = false
        private set
    var matcherSuperClassName: String? = null
        private set
    var sourceClass: PsiClass? = null
        private set
    var baseClassName: String? = null
        private set
    var factoryMethodSuffix: String? = null
        private set
    var templateFactoryMethodSuffix: String? = null
        private set
    var setterPrefix: String? = null
        private set
    var setterSuffix: String? = null
        private set
    var isUseReflectingPropertyMatcher = false
        private set
    var isGenerateTemplateFactoryMethod = false
        private set
    var isMakeMethodParametersFinal = false
        private set
    var isGenerateInterface = false
        private set

    fun setFactoryMethodPrefix(factoryMethodPrefix: String?): MatcherGeneratorProperties {
        this.factoryMethodPrefix = factoryMethodPrefix
        return self()
    }

    fun setExtensible(extensible: Boolean): MatcherGeneratorProperties {
        isExtensible = extensible
        return self()
    }

    fun setMatcherSuperClassName(matcherSuperClassName: String?): MatcherGeneratorProperties {
        this.matcherSuperClassName = matcherSuperClassName
        return self()
    }

    fun setSourceClass(sourceClass: PsiClass?): MatcherGeneratorProperties {
        this.sourceClass = sourceClass
        return self()
    }

    fun setBaseClassName(baseClassName: String?): MatcherGeneratorProperties {
        this.baseClassName = baseClassName
        return self()
    }

    fun setFactoryMethodSuffix(factoryMethodSuffix: String?): MatcherGeneratorProperties {
        this.factoryMethodSuffix = factoryMethodSuffix
        return self()
    }

    fun setTemplateFactoryMethodSuffix(templateFactoryMethodSuffix: String?): MatcherGeneratorProperties {
        this.templateFactoryMethodSuffix = templateFactoryMethodSuffix
        return self()
    }

    fun setSetterPrefix(setterPrefix: String?): MatcherGeneratorProperties {
        this.setterPrefix = setterPrefix
        return self()
    }

    fun setSetterSuffix(setterSuffix: String?): MatcherGeneratorProperties {
        this.setterSuffix = setterSuffix
        return self()
    }

    fun setUseReflectingPropertyMatcher(useReflectingPropertyMatcher: Boolean): MatcherGeneratorProperties {
        isUseReflectingPropertyMatcher = useReflectingPropertyMatcher
        return self()
    }

    fun setGenerateTemplateFactoryMethod(generateTemplateFactoryMethod: Boolean): MatcherGeneratorProperties {
        isGenerateTemplateFactoryMethod = generateTemplateFactoryMethod
        return self()
    }

    fun setMakeMethodParametersFinal(makeMethodParametersFinal: Boolean): MatcherGeneratorProperties {
        isMakeMethodParametersFinal = makeMethodParametersFinal
        return self()
    }

    fun setGenerateInterface(generateInterface: Boolean): MatcherGeneratorProperties {
        isGenerateInterface = generateInterface
        return self()
    }

    val sourceSuperClassType: PsiClassType?
        get() {
            return sourceClass?.extendsListTypes?.elementAtOrNull(0)
        }

    override val codeWriter: CodeWriter
        get() = if (isGenerateInterface) MatcherInterfaceCodeWriter(this) else MatcherClassCodeWriter(this)
}
